packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart

```diff
@@ -104,13 +123,116 @@ void main() {
+            mockLogReader.addLine('Observatory listening on http://127.0.0.1:0001');
+            mockLogReader.addLine('Observatory listening on http://127.0.0.1:1234');
+
+            time.elapse(const Duration(milliseconds: 200));
+
+            compilerStdoutController
+              .add(utf8.encode('result abc\nline1\nline2\nabc\nabc /path/to/main.dart.dill 0\n'));
+            time.flushMicrotasks();
+
+            testHotRunnerFactory.fakeAppFinishedUnexpectedly();
+            time.flushMicrotasks();
+
+            // Second iteration of the attach loop.
+            mockLogReader.addLine('Observatory listening on http://127.0.0.1:0002');
+            mockLogReader.addLine('Observatory listening on http://127.0.0.1:1235');
+
+            time.elapse(const Duration(milliseconds: 200));
+
+            compilerStdoutController
+              .add(utf8.encode('result abc\nline1\nline2\nabc\nabc /path/to/main.dart.dill 0\n'));
+            time.flushMicrotasks();
+
+            dartProcessExitCode.complete(0);
+
+            await loggerSubscription.cancel();
+            await testHotRunnerFactory.exitApp();
+            await task;
+          }));
+        });
+
+        expect(observatoryLogs.length, 7);
+        expect(observatoryLogs[0], '[stdout] Waiting for a connection from Flutter on MockAndroidDevice...');
+        expect(observatoryLogs[1], '[verbose] Observatory URL on device: http://127.0.0.1:1234');
+        expect(observatoryLogs[2], '[stdout] Lost connection to device.');
+        expect(observatoryLogs[3].contains('To hot reload changes while running, press "r". To hot restart (and rebuild state), press "R"'), isTrue);
+        expect(observatoryLogs[4], '[verbose] Observatory URL on device: http://127.0.0.1:1235');
+        expect(observatoryLogs[5], '[stdout] Lost connection to device.');
+        expect(observatoryLogs[6].contains('To hot reload changes while running, press "r". To hot restart (and rebuild state), press "R"'), isTrue);
+
+        verify(portForwarder.forward(1234, hostPort: anyNamed('hostPort'))).called(1);
+        verify(portForwarder.forward(1235, hostPort: anyNamed('hostPort'))).called(1);
+
+      }, overrides: <Type, Generator>{
+        FileSystem: () => testFileSystem,
+        HttpClientFactory: () => () => httpClient,
+        ProcessManager: () => mockProcessManager,
+        Logger: () => logger,
+        VMServiceConnector: () => getFakeVmServiceFactory(
+          vmServiceDoneCompleter: vmServiceDoneCompleter,
+        ),
+      });
+
       testUsingContext('Fails with tool exit on bad Observatory uri', () async {
         when(device.getLogReader()).thenAnswer((_) {
           // Now that the reader is used, start writing messages to it.
-          Timer.run(() {
-            mockLogReader.addLine('Foo');
-            mockLogReader.addLine('Observatory listening on http:/:/127.0.0.1:$devicePort');
-          });
+          mockLogReader.addLine('Foo');
```

Comment:
Why do some tests put this in `Timer.run()` and some don't? Also why is the `mockLogReader` disposed here, but not in other places?

Reply:
Removed the last one in attach_test.dart and protocol_discovery_test.dart.  `Timer.run` isn't needed because the throttle transformer already delays the completion of the `Future` more than one microtask.